contracts/modules/capital/CowSwapOperator.sol

```diff
@@ -0,0 +1,240 @@
+      pool.transferAssetToSwapOperator(ETH, totalOutAmount);
+      weth.deposit{value: totalOutAmount}();
+
+      // Set pool's swapValue
+      pool.setSwapValue(totalOutAmount);
+    } else if (isBuyingEth(order)) {
+      // Validate min/max setup for sellToken
+      IPool.SwapDetails memory swapDetails = pool.getAssetSwapDetails(address(order.sellToken));
+      require(swapDetails.minAmount != 0 || swapDetails.maxAmount != 0, "SwapOp: sellToken is not enabled");
+      uint256 sellTokenBalance = order.sellToken.balanceOf(address(pool));
+      require(sellTokenBalance > swapDetails.maxAmount, "SwapOp: can only sell asset when > maxAmount");
+      require(sellTokenBalance - totalOutAmount >= swapDetails.minAmount, "SwapOp: swap brings sellToken below min");
+      require(sellTokenBalance - totalOutAmount <= swapDetails.maxAmount, "SwapOp: swap leaves sellToken above max");
+
+      // Ask oracle how much ether we should get
+      uint256 oracleBuyAmount = priceFeedOracle.getEthForAsset(address(order.sellToken), order.sellAmount);
+
+      // Calculate slippage and minimum amount we should accept
+      uint256 maxSlippageAmount = (oracleBuyAmount * swapDetails.maxSlippageRatio) / MAX_SLIPPAGE_DENOMINATOR;
+      uint256 minBuyAmountOnMaxSlippage = oracleBuyAmount - maxSlippageAmount;
+      require(order.buyAmount >= minBuyAmountOnMaxSlippage, "SwapOp: order.buyAmount too low (oracle)");
+
+      // Transfer ERC20 asset from Pool
+      pool.transferAssetToSwapOperator(address(order.sellToken), totalOutAmount);
+
+      // Calculate swapValue using oracle and set it on the pool
+      uint256 swapValue = priceFeedOracle.getEthForAsset(address(order.sellToken), totalOutAmount);
+      pool.setSwapValue(swapValue);
+    } else {
+      revert("SwapOp: Must either sell or buy eth");
+    }
+
+    // Approve Cow's contract to spend sellToken
+    approveVaultRelayer(order.sellToken, totalOutAmount);
+
+    // Store the order UID
+    currentOrderUID = orderUID;
+
+    // Sign the Cow order
+    cowSettlement.setPreSignature(orderUID, true);
+
+    // Emit an event
+    emit OrderPlaced(order);
+  }
+
+  function closeOrder(GPv2Order.Data calldata order) external {
+    // Validate there is an order in place
+    require(currentOrderUID.length > 0, "SwapOp: No order in place");
+
+    // Before validTo, only controller can call this. After it, everyone can call
+    if (block.timestamp <= order.validTo) {
+      require(msg.sender == swapController, "SwapOp: only controller can execute");
+    }
+
+    validateUID(order, currentOrderUID);
+
+    // Check how much of the order was filled, and if it was fully filled
+    uint256 filledAmount = cowSettlement.filledAmount(currentOrderUID);
+    bool fullyFilled = filledAmount == order.sellAmount;
+
+    // Cancel signature and unapprove tokens
+    if (!fullyFilled) {
+      cowSettlement.setPreSignature(currentOrderUID, false);
+      approveVaultRelayer(order.sellToken, 0);
+    }
```

Comment:
Should we get rid of the check and perform these actions anyway? I'm thinking that in an extreme scenario, if the cowSettlement contract is lying (for example if it was compromised) we would want to remove the signature and reset the approval anyway - to be safe. Any counter-arguments?

Reply:
This `if` statement is for saving a moderate amount of gas, since when the order is fully filled, there's no need to execute the 2 actions.

Reply:
Perhaps the security aspect is more important than the gas savings we'd gain here.